### ISSUE TITLE
Local api server

### ISSUE THREAD

**Issue by StepanenkoNS** [tdlib/telegram-bot-api#367]
Hi, team!

I'm trying to build a backend + bot to process large (50Mb < file < 1024Mb) files in this manner:
	1) backend: send locally stored files to multiple subscribers (on schedule, one file could be sent to multiple users)
	2) bot: receive files from subscribers and store them locally for further processing.

I've got a bunch of questions and assumptions :-) Could you please be so king to clarify/confirm my understanding?

1) Sending files - process:

As far as I understand, the best way to send a file with bot is to send it using file_id. This is especially important when you want to send the same file to multiple different users.
The question is - how to obtain this file_id? I wasn't able to find any clear 'upload' methods, only 'send' methods are available.

So, I guess, there are 3 options:
	1. Use a 'system agent' user account - bot should send each file to this account beforehand and obtain file_id (local api server is required in order to bypass 50Mb limit)
	2. Vise versa - use a 'system agent' to send a file to the bot via telegram api (local api server is not required as telegram api limits are way higher than bot limits)
	3. Just send a file to the first recipient in the queue. Get file_id from this first send request and hydrate local file metadata, so the next send requests in the queue would be able to reuse this file_id instead of repetitive uploads.

Am I right? I might be missing smth...

2) Downloading large files:

	The only option to download large files that were sent to a bot is to use local api server. On receiving a 'getFile' request,  local api server will download the file from the core telegram and save it to the local storage

3) Sending files - network IO:
	
Given that I use a local api server, how does it interact with the core telegram servers while uploading/sending a file?
I guess, that file is uploaded to both (local and core) servers simultaneously or nearly simultaneously. So, network IO should be in general fileSize x2 (in and out)
Am I right?

4) Serving files

	Do I understand correctly, that local api server acts only as a proxy? So each time a bot sends a file to a client/subscriber, the file is served from the core telegram facility, not from my local api server.
	In short words, there are no additional traffic costs applied to the local api server

5) Do you have any configuration calculators/ recommendations for local api server?

     E.g. 10 bots with 1000 users each with 10 messages sent per minute will require 1Cpu + 4Gb

**Comment by levlam** [comment]
> The question is - how to obtain this file_id?

Just send the file to the first recipient in the queue and send it by file_id to all other recepients.

> So, network IO should be in general fileSize x2 (in and out)

If you use a local Bot API server, you can upload a local file, so there will be no "x2". If you need to download the file first from a remote server, or sending the request from a remote server, then the file is downloaded once to the local server and uploaded once to Telegram servers.

> In short words, there are no additional traffic costs applied to the local api server

Yes.

> E.g. 10 bots with 1000 users each with 10 messages sent per minute will require 1Cpu + 4Gb

You need at least 2 CPU cores to avoid clashing with other software and OS daemons. There are no reasons to have more than 2 cores for the server for most usages.
RAM requirements significantly depends on the number of users and chats the bots have in total. I would recommend to have at least 20 KB RAM per chat, so at least 200 MB in total for 10 bots with 1000 chats in each. 4 GB should be a safe choice for such load.
